Allow swarm join with `--availability=drain`

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #24596 where it was not possible to join as manager only (`--availability=drain`).

**The related PR in swarmkit is:**
https://github.com/docker/swarmkit/pull/1271

**\- How I did it**

This fix adds a new flag `--availability` to `swarm join`.

**\- How to verify it**

An integration test has been added.

**\- Description for the changelog**

Add a new flag `--availability` to `swarm join`.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #24596.

**The related PR in swarmkit is:**

https://github.com/docker/swarmkit/pull/1271

vdemeester commented:
/cc @thaJeztah @aaronlehmann @stevvooe @aluzzardi

thaJeztah commented:
I'm +1 on design, I think it makes sense to haven an option to set the initial availability of a node

thaJeztah commented:
Tentatively adding this to the milestone (because the related issue is), but it's no show-stopper

aaronlehmann commented on vendor/src/github.com/docker/swarmkit/api/ca.proto:
```diff
@@ -43,6 +43,9 @@ message IssueNodeCertificateRequest {
 	string token = 3;
+
+	// Availability allows a user to control the current scheduling status of a node.
+	string availability = 4;
```
This will be ignored for a renewal. I think that's okay, but we need to make it clear in the comment.

I'm not sure we want to add `availability` to the top level of `IssueNodeCertificateRequest`. Either there should be a nested messages with parameters for the new node object, or we should pass a `NodeSpec` and only honor certain fields from it (because not everything in a `NodeSpec` is under the node's control).

@aluzzardi @stevvooe any thoughts?

aluzzardi replied:
I don't like much `availability` leaking to the `NodeCA`.

Also - if we go down this path, we should probably take an initial NodeSpec rather than a bunch of fields

stevvooe replied:
This really isn't the right place for it.

Agree with the concept of providing a `NodeSpec` at join time.

aaronlehmann replied:
@yongtang: It sounds like the consensus is that we should provide a `NodeSpec` instead of just an availability setting. But as I mentioned earlier, the node CA should only use certain fields from this spec when it creates the node. There are things like node labels that are controlled from elsewhere.

stevvooe commented on vendor/src/github.com/docker/engine-api/types/swarm/swarm.go:
```diff
@@ -90,6 +90,7 @@ type JoinRequest struct {
 	AdvertiseAddr string
 	RemoteAddrs   []string
 	JoinToken     string // accept by secret
+	Availability  string
```
Surely, we have a type for this.

aaronlehmann replied:
Yes, the type is called `NodeAvailability`.

stevvooe replied:
Bump. Please use the type.

yongtang replied:
Thanks @stevvooe. The type has been changed to `NodeAvailability`.

stevvooe commented:
In general, this is a good feature, but the implementation needs some design work.

Do we have something open in swarmkit?

yongtang commented:
@stevvooe @aaronlehmann @aluzzardi @vdemeester @thaJeztah Thanks a lot for the review and help!

I was looking for feedbacks on this PR before, so I haven't created the pull request on swarmkit side yet.

Let me rework on this pull request based on the comments. Will work on `NodeSpec` and update the pull request soon.

icecrime commented:
Thanks @yongtang! In the meantime, I'll remove the milestone as it won't make it in time for 1.12.0 (which is ok).

cpuguy83 commented:
ping @stevvooe

stevvooe commented:
@cpuguy83 I am not necessarily on board with using `NodeSpec` here, since we don't honer the other aspects of the structure, such as node role.

stevvooe commented on daemon/cluster/cluster.go:
```diff
@@ -235,7 +235,7 @@ func (c *Cluster) reconnectOnFailure(n *node) {
 }
 
-func (c *Cluster) startNewNode(forceNewCluster bool, localAddr, remoteAddr, listenAddr, advertiseAddr, joinAddr, joinToken string) (*node, error) {
+func (c *Cluster) startNewNode(forceNewCluster bool, localAddr, remoteAddr, listenAddr, advertiseAddr, joinAddr, joinToken string, availability types.NodeAvailability) (*node, error) {
```
This is way too many string arguments in a row.

yongtang replied:
Thanks @stevvooe. The pull request has been updated with `localAddr`, `listenAddr`, and `advertiseAddr` into one consolidated `nodeAddr`. Please let me know if there are any issues.

cpuguy83 commented:
ping @yongtang

yongtang commented:
Thanks @cpuguy83. The pull request has been rebased and updated. Please let me know if there are any issues.

thaJeztah commented:
Sorry @yongtang this needs another rebase

ping @stevvooe is this ok to be moved to code review?

stevvooe commented:
This was my last comment here:

> @cpuguy83 I am not necessarily on board with using NodeSpec here, since we don't honer the other aspects of the structure, such as node role.

If we can reconcile this, then we can probably move along.

aaronlehmann commented:
To unblock this, maybe we can go back to the approach of not using `NodeSpec`.

Using `NodeSpec` was my suggestion. I'm sorry it delayed this so much.

stevvooe commented:
@aaronlehmann I have no problem using `NodeSpec`, as long as we don't cherry pick parameters. I think the problem is that `NodeSpec` can set things like the name, which should be under control of the cluster administrator, or membership, which is the purview of the managers. `NodeSpec` is supposed to be the user input for a node.

Let's just make this a field value for "request availability", ultimately leaving the policy to the orchestrator as to whether it is honored or not.

Do we also incorporate this into the join token?

aaronlehmann commented:
> Let's just make this a field value for "request availability", ultimately leaving the policy to the orchestrator as to whether it is honored or not.

Agreed.

> Do we also incorporate this into the join token?

My feeling is that we shouldn't, to prevent the setting from being opaque and to limit the number of permutations on the join token from exploding, but it's an interesting idea.

thaJeztah commented:
@yongtang do you have enough information to work on this?

yongtang commented:
Thanks @stevvooe @aaronlehmann @thaJeztah. Let me take a look and update the PR accordingly.

yongtang commented:
@stevvooe @aaronlehmann @thaJeztah The PR has been updated. Please take a look and let me know if there are any issues.

aaronlehmann commented on cli/command/swarm/join.go:
```diff
@@ -37,6 +39,7 @@ func newJoinCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address (format: <ip|interface>[:port])")
 	flags.StringVar(&opts.advertiseAddr, flagAdvertiseAddr, "", "Advertised address (format: <ip|interface>[:port])")
 	flags.StringVar(&opts.token, flagToken, "", "Token for entry into the swarm")
+	flags.StringVar(&opts.availability, flagAvailability, "", "Availability of the node (active/pause/drain)")
```
Should this have a default value of "active"?

yongtang replied:
Thanks @aaronlehmann. The PR has been updated with default value `active`.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -276,21 +284,31 @@ func (c *Cluster) startNewNode(forceNewCluster bool, localAddr, remoteAddr, list
 		}
 	}
 
+	specAvailability := swarmapi.NodeAvailabilityActive
```
By default, this should be an empty string. We should only send an availability value if the user specifically chooses one with the `--availability` flag.

yongtang replied:
Thanks. Done.

aaronlehmann commented on vendor/src/github.com/docker/swarmkit/ca/server.go:
```diff
@@ -233,8 +233,9 @@ func (s *Server) IssueNodeCertificate(ctx context.Context, request *api.IssueNod
-					Membership: api.NodeMembershipAccepted,
+					Role:         role,
+					Membership:   api.NodeMembershipAccepted,
+					Availability: request.Availability,
```
Before using this in `NodeSpec`, check that it is non-empty. Old engines won't include this in the request (and I think new ones shouldn't as well, unless they need to request a specific availability value).

yongtang replied:
@aaronlehmann The `request.Availability` is defined as 
```
type NodeSpec_Availability int32
```
It seems the default value will be `0` so not sure we could check if this field is set or not.

aaronlehmann replied:
Sorry, for some reason I thought this was a string. In this case, I think it makes sense to make it a pointer, and keep the value `nil` if there the user did not override it. On the gRPC side, this means splitting it out into a separate message.

yongtang replied:
@aaronlehmann Currently the definition of `Availability` is embedded inside the `NodeSpec`:
```
message NodeSpec {
        Annotations annotations = 1 [(gogoproto.nullable) = false];

        enum Membership {
                option (gogoproto.goproto_enum_prefix) = false;

                PENDING = 0 [(gogoproto.enumvalue_customname) = "NodeMembershipPending"];
                ACCEPTED = 1 [(gogoproto.enumvalue_customname) = "NodeMembershipAccepted"];
        }

        enum Availability {
                option (gogoproto.goproto_enum_prefix) = false;

                // Active nodes.
                ACTIVE = 0 [(gogoproto.enumvalue_customname) = "NodeAvailabilityActive"];
                ...
        }
```

Do we want to move `Availability` outside of `NodeSpec` (change `NodeSpec`)? Or we could spin off a new message (maybe `NodeAvailability`?) to include `NodeSpec.Availability`, and keep current `NodeSpec` message intact, like the following?
```
message NodeAvailability {

        NodeSpec.Availability availability = 1;

}
```

aaronlehmann replied:
I don't think we should move `Availability` out of `NodeSpec`.

What I was thinking of was having a wrapper mesasge instead of using `NodeSpec.Availability` directly in `IssueNodeCertificateRequest`. Something like this:

```
message NodeAvailability {
      NodeSpec.Availability availability = 1;
}
```

and in `IssueNodeCertificateRequest`:

```
...
      // Availability allows a user to control the current scheduling status of a node
      NodeAvailability availability = 4;
```

This means `availability` can be nil. If the user does not specify an availability, we can choose a default on the server side.

stevvooe commented on daemon/cluster/cluster.go:
```diff
@@ -90,6 +90,14 @@ type Config struct {
 
+// NodeAddress consists of related address of a node:
+// Local Address, Listen Address, and Advertise Address
+type NodeAddress struct {
```
Why is this type exported?

yongtang replied:
Thanks @stevvooe. The changes related to this part has already been done in PR #27698. I just removed related code accordingly.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -159,7 +167,7 @@ func New(config Config) (*Cluster, error) {
 		return nil, err
 	}
 
-	n, err := c.startNewNode(false, st.LocalAddr, st.RemoteAddr, st.ListenAddr, st.AdvertiseAddr, "", "")
```
Happy to examine it.

- LocalAddr: The node's local address. Needed for setting up a tunnel.
- RemoteAddr: The address passed of an existing node passed to `swarm join`.
- ListenAddr: The bind address. This is usually 0.0.0.0 or `::`, so it is not redundant with `LocalAddr`.
- AdvertiseAddr: Externally-reachable address that's advertised to other nodes. Needed in one-to-one NAT scenarios.

yongtang commented on cli/command/swarm/join.go:
```diff
@@ -50,6 +53,18 @@ func runJoin(dockerCli *command.DockerCli, opts joinOptions) error {
+	case swarm.NodeAvailabilityPause:
+		req.Availability = swarm.NodeAvailabilityPause
+	case swarm.NodeAvailabilityDrain:
+		req.Availability = swarm.NodeAvailabilityDrain
```
Thanks. Done.

yongtang commented:
Thanks @stevvooe @aaronlehmann @thaJeztah for the review. The PR has been updated and the comments have been addressed. (Other than the none-empty check of `request.Availability`).

Please take a look and let me know if there are any issues.

yongtang commented:
Also, the related PR in swarmkit (https://github.com/docker/swarmkit/pull/1271) has been updated as well. Please take a look and let me know for any additional issues.

aaronlehmann commented on cli/command/swarm/join.go:
```diff
@@ -50,6 +53,15 @@ func runJoin(dockerCli *command.DockerCli, opts joinOptions) error {
 		AdvertiseAddr: opts.advertiseAddr,
 		RemoteAddrs:   []string{opts.remote},
 	}
+	availability := strings.ToLower(opts.availability)
```
It might make sense to do the `swarm.NodeAvailability` cast here, to avoid the casts below.

yongtang replied:
Thanks @aaronlehmann. Done.

aaronlehmann commented on cli/command/swarm/join.go:
```diff
@@ -37,6 +39,7 @@ func newJoinCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.Var(&opts.listenAddr, flagListenAddr, "Listen address (format: <ip|interface>[:port])")
 	flags.StringVar(&opts.advertiseAddr, flagAdvertiseAddr, "", "Advertised address (format: <ip|interface>[:port])")
 	flags.StringVar(&opts.token, flagToken, "", "Token for entry into the swarm")
+	flags.StringVar(&opts.availability, flagAvailability, "active", "Availability of the node (active/pause/drain)")
```
Do you think we should support this flag on `docker swarm init` too?

aaronlehmann replied:
It's probably easiest to do it in this PR.

yongtang commented on cli/command/swarm/join.go:
```diff
@@ -50,6 +53,15 @@ func runJoin(dockerCli *command.DockerCli, opts joinOptions) error {
+	availability := strings.ToLower(opts.availability)
+	switch availability {
+	case string(swarm.NodeAvailabilityActive), string(swarm.NodeAvailabilityPause), string(swarm.NodeAvailabilityDrain):
+		req.Availability = swarm.NodeAvailability(availability)
```
Thanks. Done.

aaronlehmann commented:
LGTM (after swarmkit PR is merged)

Thanks for sticking with this!

yongtang commented:
Thanks @aaronlehmann for the review. The windowsRS1 is failing but that seems to be an Jenkins issue.

yongtang commented:
Thanks all for the review. The PR has been updated with vendoring of swarmkit done. Now all tests passes. Please take a look and let me know if there are additional issues.

aaronlehmann commented:
LGTM after rebase

yongtang commented:
@aaronlehmann Rebased. Thanks.

cpuguy83 commented:
What's the status here?

aaronlehmann commented:
The swarmkit-side PR was merged, so this is most of the way there. Sadly, it needs yet another rebase. It still looks good to me. Needs a second LGTM.

yongtang committed:
Allow swarm join with `--availability=drain`
This fix tries to address the issue raised in 24596 where it was not
possible to join as manager only (`--availability=drain`).

This fix adds a new flag `--availability` to `swarm join`.

Related documentation has been updated.

An integration test has been added.

NOTE: Additional pull request for swarmkit and engine-api will
be created separately.

This fix fixes 24596.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Allow swarm init with `--availability=drain`
This fix adds a new flag `--availability` to `swarm join`.

Related documentation has been updated.

An integration test has been added.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@cpuguy83 @aaronlehmann The PR has been rebased.

cpuguy83 reviewed (approved):
LGTM

aaronlehmann commented:
@thaJeztah: What are the next steps to get this merged? It's not an urgent PR, but I feel bad about it being open so long, and asking for rebases every few weeks.

thaJeztah commented:
oh, my bad, lost track of this one 😞

thaJeztah reviewed (changes requested):
left some suggestions

yongtang commented on docs/reference/commandline/swarm_join.md:
```diff
@@ -94,6 +95,15 @@ This flag is generally not necessary when joining an existing swarm.
+### `--availability string`
+
+This flag specifies the availability of the node at the time the node joins a master.
+Possiblle availability value includes `active`, `pause`, or `drain`.
```
Thanks. Done.

yongtang commented on docs/reference/commandline/swarm_join.md:
```diff
@@ -94,6 +95,15 @@ This flag is generally not necessary when joining an existing swarm.
 
 Secret value required for nodes to join the swarm
 
+### `--availability string`
```
Thanks. Done.

yongtang commented on docs/reference/commandline/swarm_init.md:
```diff
@@ -133,6 +134,16 @@ Snapshots compact the Raft log and allow for more efficient transfer of the
+### `--availability string`
+
+This flag specifies the availability of the node at the time the node joins a master.
+Possiblle availability value includes `active`, `pause`, or `drain`.
```
Done.

yongtang commented on docs/reference/commandline/swarm_init.md:
```diff
@@ -133,6 +134,16 @@ Snapshots compact the Raft log and allow for more efficient transfer of the
 state to new managers. However, there is a performance cost to taking snapshots
 frequently.
 
+### `--availability string`
```
Done.

yongtang commented:
Thanks @aaronlehmann @thaJeztah the PR has been updated.

thaJeztah reviewed (approved):
LGTM

genki commented:
Is there a way to become availability=active from worker node itself?
